src/compiler/scala/tools/nsc/ast/parser/Scanners.scala

```diff
@@ -543,24 +543,40 @@ trait Scanners extends ScannersCommon {
           }
           fetchDoubleQuote()
         case '\'' =>
+          def unclosedCharLit() = {
+            val unclosed = "unclosed character literal"
+            // advise if previous token was Symbol contiguous with the orphan single quote at offset
+            val msg =
+              if (token == SYMBOLLIT && offset == lastOffset) s"""$unclosed (or use " for string literal "$strVal")"""
+              else unclosed
+            syntaxError(msg)
+          }
           def fetchSingleQuote() = {
             nextChar()
             if (isIdentifierStart(ch))
               charLitOr(getIdentRest)
             else if (isOperatorPart(ch) && (ch != '\\'))
               charLitOr(getOperatorRest)
             else if (!isAtEnd && (ch != SU && ch != CR && ch != LF || isUnicodeEscape)) {
+              val isEmptyCharLit = (ch == '\'')
               getLitChar()
               if (ch == '\'') {
-                nextChar()
-                token = CHARLIT
-                setStrVal()
+                if (isEmptyCharLit && settings.isScala213)
```

Comment:
Sadly, the `isScala213` check is expensive (`def isScala213: Boolean = source.value >= version213`). This keeps coming up, maybe it's time to do something about this (/cc performance team @retronym, @lrytz).